activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -9,15 +9,14 @@ class << self
         def included(base) #:nodoc:
           dirties_query_cache base, :insert, :update, :delete, :rollback_to_savepoint, :rollback_db_transaction
 
-          base.set_callback :checkout, :after, :configure_query_cache!
           base.set_callback :checkin, :after, :disable_query_cache!
```

Comment:
The query cache state for all connections and pools is now tracked on the handler, but this callback still disables the cache. This means that, for example, checking in _any_ connection will disable the cache for the entire thread:

```ruby
  def test_query_cache_state_is_unaffected_when_another_connection_is_checked_in
    with_temporary_connection_pool do
      middleware {
        ActiveRecord::Base.connection_pool.with_connection { }
        assert_cache :clean
      }.call({})
    end
  end
```

This should probably call `clear_query_cache` now instead.

Reply:
I've been debugging this the last day or so and I'm not sure how to fix this. Changing `disable_query_cache!` to `clear_query_cache` fixes _this_ test but it breaks the test I added and pushed up. 

Mainly, clear_active_connections _should_ disable the cache on all connections, but by changing this to clear_query_cache this test will fail. I've pushed it up with your suggested change so you can see what I mean. Do you have any ideas about how to resolve this?

```ruby
  test "query cache on clear_active_connections" do
    with_temporary_connection_pool do
      ActiveRecord::Base.connection_pool.connections.each do |conn|
        assert_cache :off, conn
      end 

      assert_not_predicate ActiveRecord::Base.connection, :nil?
      assert_cache :off

      middleware do |env|
        assert_cache :clean

        Task.find 1
        assert_cache :dirty

        ActiveRecord::Base.clear_active_connections!

        ActiveRecord::Base.connection_pool.connections.each do |conn|
          assert_cache :off, conn
        end 
      end.call({})
    end 
  end 
```

Reply:
Calling `disable_query_cache!` in `clear_active_connections!` makes your new test pass:

```patch
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
index 0168afcf8c..9908f0da0f 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
@@ -1003,6 +1003,7 @@ def active_connections?
       # and also returns connections to the pool cached by threads that are no
       # longer alive.
       def clear_active_connections!
+        disable_query_cache!
         connection_pool_list.each(&:release_connection)
       end
```

I guess `clear_reloadable_connections!` and `clear_all_connections!` should do the same?

Reply:
Here's another interesting case:

```ruby
  test "query cache doesn't persist while the cache is disabled" do
    with_temporary_connection_pool do
      middleware {
        other_connection = ActiveRecord::Base.connection_pool.checkout
        assert_equal 0, Post.where(title: "foo").count

        other_connection.disable_query_cache!
        Post.create!(title: "foo", body: "bar")

        other_connection.enable_query_cache!
        assert_equal 1, Post.where(title: "foo").count
      }.call({})
    end
  end
```

I think there's two issues here:
 - Calling `connection.disable_query_cache!` seems like it should only affect that connection, but now it disables the cache globally.
 - Disabling the query cache on the handler doesn't clear the cached results for each connection, which can then become stale and be reused later if the cache is reenabled.